ecosystem/indexer-grpc/indexer-grpc-data-service/src/service.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_indexer_grpc_utils::{
+    get_cache_coverage_status, get_cache_transactions, get_file_store_bucket_name,
+    get_redis_address,
+    storage::{generate_blob_name, TransactionsBlob, BLOB_TRANSACTION_CHUNK_SIZE},
+    CacheCoverageStatus,
+};
+use aptos_logger::{info, warn};
+use aptos_moving_average::MovingAverage;
+use aptos_protos::datastream::v1::{
+    indexer_stream_server::IndexerStream,
+    raw_datastream_response::Response as DatastreamProtoResponse, RawDatastreamRequest,
+    RawDatastreamResponse, TransactionOutput, TransactionsOutput,
+};
+use cloud_storage::Object;
+use futures::Stream;
+use redis::{Client, Commands};
+use std::{pin::Pin, thread::sleep, time::Duration};
+use tokio::sync::mpsc::{self, error::TrySendError};
+use tokio_stream::wrappers::ReceiverStream;
+use tonic::{Request, Response, Status};
+use uuid::Uuid;
+
+type ResponseStream = Pin<Box<dyn Stream<Item = Result<RawDatastreamResponse, Status>> + Send>>;
+
+pub struct DatastreamServer {
+    pub redis_client: Client,
+}
+
+impl DatastreamServer {
+    pub fn new() -> Self {
+        Self {
+            redis_client: Client::open(format!("redis://{}", get_redis_address())).unwrap(),
+        }
+    }
+}
+
+impl Default for DatastreamServer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+// DatastreamServer handles the raw datastream requests from cache and file store.
+#[tonic::async_trait]
+impl IndexerStream for DatastreamServer {
+    type RawDatastreamStream = ResponseStream;
+
+    async fn raw_datastream(
+        &self,
+        req: Request<RawDatastreamRequest>,
+    ) -> Result<Response<Self::RawDatastreamStream>, Status> {
+        let (tx, rx) = mpsc::channel(10000);
+
+        let mut conn = self.redis_client.get_connection().unwrap();
+
+        let req = req.into_inner();
+        // Round the version to the nearest STORAGE_BLOB_SIZE.
+        let mut current_version =
+            req.starting_version / BLOB_TRANSACTION_CHUNK_SIZE * BLOB_TRANSACTION_CHUNK_SIZE;
+
+        tokio::spawn(async move {
+            let mut ma = MovingAverage::new(10_000);
```

Comment:
There's a lot of magic numbers, let's move these to constants

Reply:
done.